backend/result_publisher.py

```diff
@@ -14,42 +14,77 @@ def __init__(self, eventListener):
         self._update_rate = 1.0
         self.socket = None
         self._last_count = 0
-
+	self._portList = ports.result_stream
+	self._portDict = {}
+	self._socketDict = {} 
+	self.default_port = self._portList[0]
+	self.histogramNum = 0
+    
     def run(self):
         log.info("Starting ResultPublisher")
-        self.connect()
-
-        self._publish_clear()
         while True:
-            count = len(self.eventListener._gather_histogram_transition.get_checkpoint())
-            if count != self._last_count:
-                self._publish_clear()
-                self._last_count = count
-            for i in range(count):
-                if self.eventListener._gather_histogram_transition.get_checkpoint()[i]:
-                    self._publish(i)
-            time.sleep(self.update_rate)
-
-    def connect(self):
-        context = zmq.Context()
-        self.socket = context.socket(zmq.PUB)
-        uri = 'tcp://*:{0:d}'.format(ports.result_stream)
-        self.socket.bind(uri)
-        log.info('Bound to ' + uri)
+	    if len(self.eventListener.transition_objects_dict['GatherHistogram']) >=1:
+		for gather_histogram_transition in self.eventListener.transition_objects_dict['GatherHistogram']:
+	     	    if self.default_port not in self._portDict.values():
+			self._portDict[gather_histogram_transition] = self.default_port
+			self.new_connection(self.default_port, gather_histogram_transition)
+		    else: 
+			if self._portDict.get(gather_histogram_transition) == None:
+			    for port in self._portList:
+				if port not in self._portDict.values():
+				    self._portDict[gather_histogram_transition] = port
+				    log.info( "Adding " + gather_histogram_transition.get_name() + " to port " + str(port))
+				    self.new_connection(port, gather_histogram_transition)
+				    break
```

Comment:
Can't this be shortened a bit by dropping the `if self.default_port not in self._portDict.values():`? It seems it could probably be covered by the `else` (potentially with small modifications). That would make things more readable, right now this is quite hard.